cip/CIP2017-06-18-multiple-graphs.adoc

```diff
@@ -0,0 +1,476 @@
+
+==== Graph URLs
+
+The exact shape and form of graph URL lies outside the scope of this CIP.
+
+This CIP however proposes that a `<graph-url>` must always be given as either a string literal or a query parameter.
+
+This allows parameterization of queries by controlling which graphs from which graph URLs they should use.
+
+=== Changing back to no graph
+
+Additionally, this CIP proposes new syntax for discarding the source and the target graph of the current query:
+
+[source, cypher]
+----
+FROM -
+INTO -
+----
+
+`-` is not a graph specifier; rather this syntax is a special form for discarding the current source and target graph such that the provided source and target graph are again chosen to be the default graph as specified for partial query contexts.
+
+In consequence, both `FROM -` and `INTO -` will not bind the default graph to a generated fresh name.
+This is different from `<graph-specifier>` semantics that will ensure that referenced graphs are always bound to a name.
+
+=== Returning, aliasing, and selecting graphs
+
+This CIP proposes to extend both the `WITH` and the `RETURN` clauses with new syntax for controlling the set of available named graphs that should be passed on by the clause (or returned from the query respectively) by explicitly specifying all `<graph-return-items>`.
+The newly proposed syntax is:
+
+[source, cypher]
+----
+WITH [ < return-items > ] [ [ INPUT ] GRAPHS < graph-return-items > ]
+RETURN [ < return-items > ] [ [ INPUT ] GRAPHS < graph-return-items > ]
+----
+
+This CIP proposes the following kinds of `<graph-return-items>`:
+
+* `<graph-specifier-list>`: A comma separated list of `<graph-specifier>` that are to be passed on
+* `*`: All named graphs are to be passed on
+* `*, <graph-specifier-list>`: All named graphs are to be passed on together with any additional named graphs that are newly bound in `<graph-specifier-list>`
+* `-`: No named graphs are to be passed on
+
+The order of named graphs inherently given by `<graph-return-items>` is semantically insignificant.
+However it is recommended that conforming implementations preserve this order at least in programmatic output operations (e.g. a textual display of the list of returned graphs).
+This in essence mirrors the semantics for tabular data returned by Cypher.
+
+Both `WITH ... GRAPHS ...` and `RETURN ... GRAPHS ...` will pass on (or return respectively) exactly the set of described named graphs.
+To simplify passing on available graphs it is proposed by this CIP that regular `WITH <return-items>` is taken to be syntactic sugar for `WITH <return-items> GRAPHS -` and that regular `RETURN <return-items>` is taken to be syntactic sugar for `RETURN <return-items> GRAPHS -`.
+
+To even further simplify, it is additionally proposed that `WITH|RETURN <return-items> INPUT GRAPHS <graph-return-items>` is to be syntactic sugar for `WITH|RETURN <return-items> GRAPHS <graph-return-items>, SOURCE GRAPH, TARGET GRAPH`.
```

Comment:
I'm not convinced of the usefulness of this syntactic sugar -- I find that it is hard to know what kind of queries will be prominent in this new model. In general, I think that it would be useful to have a little less focus on the syntactic sugar bits, and more on the core model. Syntactic sugar additions could always follow later.